modules/images/image-loading-optimization/class-ilo-html-tag-processor.php

```diff
@@ -226,21 +226,30 @@ public function open_tags(): Generator {
 	/**
 	 * Gets breadcrumbs for the current open tag.
 	 *
-	 * Breadcrumbs are constructed to match the format from detect.js.
+	 * A breadcrumb consists of a tag name and its sibling index.
 	 *
-	 * TODO: Consider rather each breadcrumb being a (tag,index) tuple.
-	 *
-	 * @return array<array{tag: string, index: int}> Breadcrumbs.
+	 * @return Generator<array{string, int}> Breadcrumb.
 	 */
-	public function get_breadcrumbs(): array {
-		$breadcrumbs = array();
+	private function get_breadcrumbs(): Generator {
 		foreach ( $this->open_stack_tags as $i => $breadcrumb_tag_name ) {
-			$breadcrumbs[] = array(
-				'tag'   => $breadcrumb_tag_name,
-				'index' => $this->open_stack_indices[ $i ],
-			);
+			yield array( $breadcrumb_tag_name, $this->open_stack_indices[ $i ] );
+		}
+	}
+
+	/**
+	 * Gets XPath for the current open tag.
+	 *
+	 * It would be nicer if this were like `/html[1]/body[2]` but in XPath the position() here refers to the
+	 * index of the preceding node set. So it has to rather be written `/*[1][self::html]/*[2][self::body]`.
```

Comment:
This being said, the breadcrumbs could be updated to keep track of the index of the tag of a given type rather than the index of the tag in general. This would make breadcrumbs more in line with the natural XPath syntax, so instead `/*[1][self::html]/*[2][self::body]` we could do just `/html[1]/body[1]`.

Nevertheless, what is implemented now is what the `:nth-child()` pseudo-selector implements in CSS, as opposed to `:nth-of-type()`.

But doing this would take some extra bookkeeping, and I don't see a clear benefit.

I also just came across a Chromium doc that is looking at the same problem space: [Identifying element consistently across reloads](https://docs.google.com/document/d/18QpJB1q7eSgzG7n3SQPC61kn5UJGaS1ZTG0Y1eaCq5w/edit).